staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go

```diff
@@ -228,14 +236,30 @@ func (o *ExposeServiceOptions) RunExpose(cmd *cobra.Command, args []string) erro
 		return cmdutil.UsageErrorf(cmd, err.Error())
 	}
 
-	// Get the generator, setup and validate all required parameters
-	generatorName := cmdutil.GetFlagString(cmd, "generator")
-	generators := o.Generators("expose")
-	generator, found := generators[generatorName]
-	if !found {
-		return cmdutil.UsageErrorf(cmd, "generator %q not found.", generatorName)
+	names := []Param{
+		{Name: "default-name", Required: true},
+		{Name: "name", Required: false},
+		{Name: "selector", Required: true},
+		// port will be used if a user specifies --port OR the exposed object
+		// has one port
+		{Name: "port", Required: false},
+		// ports will be used iff a user doesn't specify --port AND the
+		// exposed object has multiple ports
+		{Name: "ports", Required: false},
+		{Name: "labels", Required: false},
+		{Name: "external-ip", Required: false},
+		{Name: "load-balancer-ip", Required: false},
+		{Name: "type", Required: false},
+		{Name: "protocol", Required: false},
+		// protocols will be used to keep port-protocol mapping derived from
+		// exposed object
+		{Name: "protocols", Required: false},
+		{Name: "container-port", Required: false}, // alias of target-port
+		{Name: "target-port", Required: false},
+		{Name: "port-name", Required: false},
+		{Name: "session-affinity", Required: false},
+		{Name: "cluster-ip", Required: false},
```

Comment:
Hi @rikatz , I was wondering if `portName` option is needed there doesn't exists `cmd.Flags()` for it before and check at line '440` was leaving it unnamed always . So should I add that option to be taken as input or remove it completely

Reply:
@rikatz , so I checked , in RunExpose(https://github.com/kubernetes/kubernetes/blob/5b0a2c3a29f6b5392e0f8f94ba5669bdc9eb73f6/staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go#L217) it wasn't being used to set the `port-name` because every time when it's generated over here (https://github.com/kubernetes/kubernetes/blob/5b0a2c3a29f6b5392e0f8f94ba5669bdc9eb73f6/staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go#L39). The name is always set to `default`. and when the generateService is called the `name` (https://github.com/kubernetes/kubernetes/blob/5b0a2c3a29f6b5392e0f8f94ba5669bdc9eb73f6/staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go#L107) over here will basically take care of setting the name, so I think we can safely remove that

Reply:
@rikatz , so I checked, in RunExpose https://github.com/kubernetes/kubernetes/blob/5b0a2c3a29f6b5392e0f8f94ba5669bdc9eb73f6/staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go#L217
there is no flag called `-port-name`. and over here is where the `name` will be parsed from cmd https://github.com/kubernetes/kubernetes/blob/5b0a2c3a29f6b5392e0f8f94ba5669bdc9eb73f6/staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go#L107, However, because `--port-name` is never set so over here it will just be empty string https://github.com/kubernetes/kubernetes/blob/5b0a2c3a29f6b5392e0f8f94ba5669bdc9eb73f6/staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go#L121 and eventually the name will be empty string over here https://github.com/kubernetes/kubernetes/blob/5b0a2c3a29f6b5392e0f8f94ba5669bdc9eb73f6/staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go#L150. 
I am guessing it's a bug?? because `ServicePort` name will never be set

Reply:
I've seen in other part of the code (need to remember), but I thought port-name actually is used on the scenario where you have a Deployment, and runs expose against it.

But I did an experiment here with the following deployment:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: nginx
  name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
        ports:
        - containerPort: 8080
          name: http
        - containerPort: 8081
          name: http2
```

and when running `kubectl expose deploy/nginx` the service gets created as the following:
```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: nginx
  name: nginx
  namespace: default
spec:
  clusterIP: 10.124.4.183
  ports:
  - name: port-1
    port: 8080
    protocol: TCP
    targetPort: 8080
  - name: port-2
    port: 8081
    protocol: TCP
    targetPort: 8081
  selector:
    app: nginx
  type: ClusterIP
```
So I don't know if this is a bug and the port names should have the name described on the deployment manifest, or if we should remove the port-names :)

Anyway, let's first move with the most complex part: removing the generators ;)